source/src/enet_encap/encap.c

```diff
@@ -606,7 +609,12 @@ int_fast32_t CreateEncapsulationStructure(const EipUint8 *receive_buffer, int re
   receive_buffer += kSenderContextSize;
   encapsulation_data->options = GetUdintFromMessage(&receive_buffer);
   encapsulation_data->current_communication_buffer_position = (EipUint8*) receive_buffer;
-  return (receive_buffer_length - ENCAPSULATION_HEADER_LENGTH - encapsulation_data->data_length);
+
+  /* Ensure buffer length fits in an int32 before casting in the return expression. */
+  OPENER_ASSERT(INT32_MAX >= receive_buffer_length);
```

Comment:
based on my previous experience with security issues from NIST, Claroty, and Cisco Talos, I would prefer not to have an ASSERT here, as this is then may open a way to kill OpENer with a malformed packet. I would prefer a way to make this impossible in the first place

Reply:
Agree, malformed data received over the network should not be fatal. I reserve assertions for tests that indicate an outright invalid condition that should never happen, regardless of system input, and execution should not continue. In that respect, it's no different than the encapsulation_data pointer; the program will crash if passed a NULL; an non-null assertion would make sense on that parameter.

At any rate, it looks like the buffer length ultimately originates from a socket call return value, not received data. The only way I can think this would happen is a huge TCP read with PC_OPENER_ETHERNET_BUFFER_SIZE > INT32_MAX; a UDP datagram that big is impossible. It's probably safe to call foul if a TCP socket read returns 2GB.

Another aspect is the original implementation is doing the following implicit conversion when calling CreateEncapsulationStructure:
```
int receive_buffer_length; // signed
size_t original_value; // unsigned
receive_buffer_length = original_value;
```
The result is implementation-defined if the `receive_buffer_length` can not represent `original_value` because they are different signedness. An assertion is probably preferable to implementation-defined weirdness.

Reply:
I agree, and theoretically packets with a data size of more than PC_OPENER_ETHERNET_BUFFER_SIZE - 4 should never get there (see generic_networkhandler.c:745).
A thought on the ASSERT condition, the only guarantee on size_t that I found was, that it will be always at least 16 bits wide. Perhaps I am wrong on that, but on such systems your condition could never be true.

I think that should not cause an issue then, but would like to hear your opinion on that too

Reply:
One way to think about it is the return expression:

`
return receive_buffer_length - ENCAPSULATION_HEADER_LENGTH - encapsulation_data->data_length;
`

The cast has been removed for clarity, as it doesn't affect the outcome; it only serves to suppress the compiler warning. We can also disregard `ENCAPSULATION_HEADER_LENGTH` and `encapsulation_data->data_length` as neither are significant enough to ensure the result will always fit into an `int32_t` for any value of `receive_buffer_length`. So, basically, it comes down to `return receive_buffer_length`.

This will result in an implicit integer conversion, which follows [these rules](https://en.cppreference.com/w/c/language/conversion):

> A value of any integer type can be implicitly converted to any other integer type. Except where covered by promotions and boolean conversions above, the rules are:
>
> -  if the target type can represent the value, the value is unchanged
>
 > - otherwise, if the target type is unsigned, the value 2b, where b is the number of value bits in the target type, is repeatedly subtracted or added to the source value until the result fits in the target type. In other words, unsigned integers implement modulo arithmetic.
>
> - otherwise, if the target type is signed, the behavior is implementation-defined (which may include raising a signal) 

The first rule will apply in the case where `size_t` is 16-bit; the assertion will never be false; no problem there. The third case applies for all other systems because `size_t` is always unsigned and `int32_t` is signed, and _implementation-defined_ isn't something to rely upon. For example, it truncates on my Linux x64 gcc system. Regardless of the implementation, _the result will be incorrect_, and will likely cause incorrect operation or an outright crash. I think there is a good argument that a `receive_buffer_length` exceeding `INT32_MAX` is almost certianly the result of a bug, and can't be handled anyway, so better to halt the program as close to the source of the problem as posssible, giving the programmer a good chance to locate the issue, as opposed to letting the program continue in an undefined manner.

Reply:
> A thought on the ASSERT condition, the only guarantee on size_t that I found was, that it will be always at least 16 bits wide. Perhaps I am wrong on that, but on such systems your condition could never be true.

I think you mean the the assert condition can never be _false_ if `size_t` is 16-bit, which is fine because `int32_t` can handle any `size_t` value in that case.

Reply:
Yes, exactly that I meant. Great, I had the same conclusion, just wanted to be sure before merging.

Thanks